### PR TITLE
Allow VB Shared New to initialize a Shared field of another instance.

### LINKED ISSUE #1028: Breaking change: VB can no longer assign to static readonly field of another instantiation
The following code compiled and ran on previous versions of VB. In Roslyn this does not compile:

```vb
Class Foo(Of T)
    Shared Sub New()
        Foo(Of Integer).X = 12
    End Sub

    Public Shared ReadOnly X As Integer
End Class
```

The native compiler allowed this code, and the code runs “correctly” when the verifier is disabled. However, it is certainly not allowed by the intent of the language specification, the code almost certainly contains an error, and the generated code violates the CLI specification (for the init-only constraint). The verifier rejects the generated code, but most invocations of the CLR skip the verifier.

I propose we continue to reject this code in Roslyn and document the incompatibility that results from the fact that the implementation has been tightened.

See also #990 

@ljw1004 @AlekseyTs @AnthonyDGreen What do you think?

### PR DESCRIPTION
**Pull request by gafter**
Fixes #1028

### CODE REVIEW

**Comment by gafter** [comment]
@jaredpar @VSadov @AlekseyTs @agocke @VladimirReshetnikov Could a couple of you review this please?

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/FieldInitializerBindingTests.vb [hunk 1]
```diff
@@ -1408,6 +1408,77 @@ BC30799: Field 'Clazz.b' has an invalid constant value.
 </errors>)
         End Sub
 
+        <Fact, WorkItem(1028, "https://github.com/dotnet/roslyn/issues/1028")>
+        Sub WriteOfReadonlyStaticMemberOfAnotherInstatiation01()
```
Comment (VladimirReshetnikov) [inline src/Compilers/VisualBasic/Test/Semantic/Semantics/FieldInitializerBindingTests.vb]: Spelling: Instatiation -> Instantiation

Comment (VladimirReshetnikov) [inline src/Compilers/VisualBasic/Test/Semantic/Semantics/FieldInitializerBindingTests.vb]: 'Static' has a different meaning in VB. Consider renaming to `WriteOfReadonlySharedMemberOfAnotherInstantiation01`

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/FieldInitializerBindingTests.vb [hunk 2]
```diff
@@ -1408,6 +1408,77 @@ BC30799: Field 'Clazz.b' has an invalid constant value.
...
+Class Foo(Of T)
+    Shared Sub New()
+        Foo(Of Integer).X = 12
+        Foo(Of Integer).Y = 12
+        Foo(Of T).X = 12
+        Foo(Of T).Y = 12
+    End Sub
+
+    Public Shared ReadOnly X As Integer
+    Public Shared ReadOnly Property Y As Integer = Nothing
```
Comment (VladimirReshetnikov) [inline src/Compilers/VisualBasic/Test/Semantic/Semantics/FieldInitializerBindingTests.vb]: Consider replacing `Nothing` with `0`.

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/FieldInitializerBindingTests.vb [hunk 3]
```diff
@@ -1408,6 +1408,77 @@ BC30799: Field 'Clazz.b' has an invalid constant value.
...
+        Foo(Of Integer).X = 12
+        ~~~~~~~~~~~~~~~~~
+BC30526: Property 'Y' is 'ReadOnly'.
+        Foo(Of Integer).Y = 12
+        ~~~~~~~~~~~~~~~~~~~~~~
+</expected>)
+        End Sub
+
+        <Fact, WorkItem(1028, "https://github.com/dotnet/roslyn/issues/1028")>
+        Public Sub WriteOfReadonlyStaticMemberOfAnotherInstatiation02()
```
Comment (VladimirReshetnikov) [inline src/Compilers/VisualBasic/Test/Semantic/Semantics/FieldInitializerBindingTests.vb]: The same comments for this name.

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/FieldInitializerBindingTests.vb [hunk 4]
```diff
@@ -1408,6 +1408,77 @@ BC30799: Field 'Clazz.b' has an invalid constant value.
...
+
+            CompilationUtils.AssertTheseDiagnostics(compilation, <expected>
+BC30526: Property 'Y' is 'ReadOnly'.
+        Foo(Of Integer).Y = 12
+        ~~~~~~~~~~~~~~~~~~~~~~
+</expected>)
+            CompilationUtils.AssertTheseDiagnostics(compilation.WithStrictMode(), <expected>
+BC30064: 'ReadOnly' variable cannot be the target of an assignment.
+        Foo(Of Integer).X = 12
+        ~~~~~~~~~~~~~~~~~
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/Semantics/FieldInitializerBindingTests.vb]: Is this error expected in a non-strict mode?

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/FieldInitializerBindingTests.vb [hunk 5]
```diff
@@ -1408,6 +1408,77 @@ BC30799: Field 'Clazz.b' has an invalid constant value.
...
+End Class
+        </file>
+   </compilation>, TestOptions.ReleaseDll)
+
+            CompilationUtils.AssertTheseDiagnostics(compilation, <expected>
+BC30526: Property 'Y' is 'ReadOnly'.
+        Foo(Of Integer).Y = 12
+        ~~~~~~~~~~~~~~~~~~~~~~
+</expected>)
+            CompilationUtils.AssertTheseDiagnostics(compilation.WithStrictMode(), <expected>
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/Semantics/FieldInitializerBindingTests.vb]: Consider adding an empty line before this one to improve readability.

**Comment by AlekseyTs** [comment]
LGTM

**Pull request closed without merging**